### PR TITLE
HPCC-22285 Execute WU Analysis post execution and write results to WU

### PR DESCRIPTION
**Pull request by shamser**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

### CODE REVIEW

**Comment by shamser** [comment]
@ghalliday

File: ecl/eclagent/eclagent.cpp [hunk 1]
```diff
@@ -1950,6 +1951,12 @@ void EclAgent::doProcess()
             break;
         }
 
+        if (w->getState() == WUStateCompleted)
+        {
+            const char * clustername = w->queryClusterName();
+            if (getClusterTypeByClusterName(clustername)==ThorLCRCluster)
+                analyseWorkunit(static_cast<IConstWorkUnit*>(w.get()));
```
Comment (ghalliday) [inline ecl/eclagent/eclagent.cpp]: Minor: would be better to pass in w as an IWorkUnit * - then there would be no need for the code to lock the workunit.  (There also shouldn't need to be a cast to convert a IWorkUnit * to a IConstWorkUnit *.

File: ecl/eclagent/eclagent.cpp [hunk 2]
```diff
@@ -1950,6 +1951,12 @@ void EclAgent::doProcess()
             break;
         }
 
+        if (w->getState() == WUStateCompleted)
+        {
+            const char * clustername = w->queryClusterName();
+            if (getClusterTypeByClusterName(clustername)==ThorLCRCluster)
```
Comment (ghalliday) [inline ecl/eclagent/eclagent.cpp]: trivial: possibly better to check clusterType - may have less overhead.

File: ecl/eclagent/eclagent.cpp [hunk 3]
```diff
@@ -1950,6 +1951,12 @@ void EclAgent::doProcess()
             break;
         }
 
+        if (w->getState() == WUStateCompleted)
```
Comment (ghalliday) [inline ecl/eclagent/eclagent.cpp]: We should have a #option that can be used to disable this code - just in case there is a problem with it that causes eclagent to crash.

**Review by ghalliday — COMMENTED** [review]
@shamser a few minor comments.  Also please rebase once the other PR is merged.

**Comment by shamser** [comment]
@ghalliday A new ticket has been created to document the configuration and #option (HPCC-22291)  Please can you review.

File: ecl/eclagent/eclagent.cpp [hunk 4]
```diff
@@ -1950,6 +1951,25 @@ void EclAgent::doProcess()
...
         }
 
+        if (w->getState() == WUStateCompleted && getClusterType(clusterType)==ThorLCRCluster)
+        {
+            if (agentTopology->getPropBool("@analyzeWorkunit", true) && w->getDebugValueBool("analyzeWorkunit", true))
+            {
+                WuAnalyseOptions options;
+                CIArrayOf<PerformanceIssue> issues;
+                analyseWorkunit(w.get(), options, issues);
+                ForEachItemIn(i, issues)
```
Comment (ghalliday) [inline ecl/eclagent/eclagent.cpp]: I preferred the structure of the previous commit.  You are breaking encapsulation of the analysis logic - previously these classes were not part of the public interface (and they are not exported from the dll).  It also means you have lost the option to run the analyser from the wutool and write the results to the workunit.  I can imagine that being useful - e.g., if you want to run new rules on an old workunit.

File: common/wuanalysis/anacommon.hpp [hunk 5]
```diff
@@ -19,6 +19,7 @@
 #define ANACOMMON_HPP
 
 #include "jliball.hpp"
+#include "workunit.hpp"
```
Comment (ghalliday) [inline common/wuanalysis/anacommon.hpp]: Why is this needed?

File: common/wuanalysis/anacommon.hpp [hunk 6]
```diff
@@ -44,4 +46,11 @@ class PerformanceIssue : public CInterface
 
 extern int compareIssuesCostOrder(CInterface * const * _l, CInterface * const * _r);
 
+struct WuAnalyseOptions
```
Comment (ghalliday) [inline common/wuanalysis/anacommon.hpp]: This was original intended to be a private include file - this should be in anawu.hpp which is the public interface to the analyser

File: ecl/eclagent/eclagent.cpp [hunk 7]
```diff
@@ -1950,6 +1951,25 @@ void EclAgent::doProcess()
             break;
         }
 
+        if (w->getState() == WUStateCompleted && getClusterType(clusterType)==ThorLCRCluster)
+        {
+            if (agentTopology->getPropBool("@analyzeWorkunit", true) && w->getDebugValueBool("analyzeWorkunit", true))
```
Comment (ghalliday) [inline ecl/eclagent/eclagent.cpp]: The following is better - since it allows explicit overriding both ways:

`(w->getDebugValueBool("analyzeWorkunit", agentTopology->getPropBool("@analyzeWorkunit", true))
`

**Review by ghalliday — COMMENTED** [review]
@shamser see some comments

File: ecl/eclagent/eclagent.cpp (hunk 4 above)
Comment (shamser) [reply ecl/eclagent/eclagent.cpp]: My thinking was that analyseWorkunit would do only that - analyse.  Updating the workunit or printing out the results doesn't seem part of the analysis. We could provide a separate function to write the issues to the WU and another function to print out the issues.

The other question is that if we were to do it the way it was done before, would we pass a locked or an unlocked workunit?  A locked workunit is not necessary for Analyze & Print but an locked workunit is required for Analyse & Update WU.  It seems that we'd need to either lock WU unnessary or need to relock the WU (that may already be locked).

File: common/wuanalysis/anacommon.hpp (hunk 5 above)
Comment (shamser) [reply common/wuanalysis/anacommon.hpp]: It doesn't need to be here - I'm removing.

File: common/wuanalysis/anacommon.hpp (hunk 6 above)
Comment (shamser) [reply common/wuanalysis/anacommon.hpp]: I understand. Will move public interface to anawu.hpp.

File: ecl/eclagent/eclagent.cpp (hunk 7 above)
Comment (shamser) [reply ecl/eclagent/eclagent.cpp]: Neat. That works better.

File: ecl/eclagent/eclagent.cpp (hunk 4 above)
Comment (ghalliday) [reply ecl/eclagent/eclagent.cpp]: There is always a balancing act for where code goes - but generally it is best to try and encapsulate code to minimize information leaking.  For instance if the results of the analysis were then going to be processed by multiple other stages then it would make sense to export the information about the results.  If (as in this case) the results have little further processing it makes sense to keep the details hidden.

As for the details, in this case I would have had two functions, one that took a IWorkUnit and added the results to the workunit exceptions, the other than took an IConstWorkUnit (which is a base interface of IWorkUnit) and printed them.  They would use common code within the anawu module.

Slightly related one rule I have found useful is that if most of the code within a function is accessing members of an object that as a good indication that should be a member function.  Which would argue against removing PerformanceIssue::print() - and similarly the code to add an issue as a workunit exception would be cleaner as a member function.

File: common/wuanalysis/anawu.cpp [hunk 8]
```diff
@@ -354,14 +352,6 @@ void WorkunitAnalyser::collateWorkunitStats(IConstWorkUnit * workunit, const WuS
     }
 }
 
-void WorkunitAnalyser::printWarnings()
```
Comment (ghalliday) [inline common/wuanalysis/anawu.cpp]: Better encapsulation to  keep this function and also add a createExceptions(wu) member, and avoids the refeference parameter on the constructor.

**Comment by ghalliday** [comment]
@shamser looks better.  One comment - I don't mind if you make the change or not.  Please squash into a single commit, and I'll merge.

**Comment by shamser** [comment]
@ghalliday

**Comment by ghalliday** [comment]
Great, looks clean.
